cpp/src/arrow/table.cc

```diff
@@ -389,7 +389,7 @@ Status Table::FromRecordBatches(const std::shared_ptr<Schema>& schema,
   const int ncolumns = static_cast<int>(schema->num_fields());
 
   for (int i = 1; i < nbatches; ++i) {
-    if (!batches[i]->schema()->Equals(*schema)) {
+    if (!batches[i]->schema()->Equals(*schema, false)) {
```

Comment:
Because of the following failing test case:

```py
pyarrow/tests/test_parquet.py:614: AssertionError
___________________________________________ test_recordbatchlist_to_pandas ____________________________________________

    def test_recordbatchlist_to_pandas():
        data1 = pd.DataFrame({
            'c1': np.array([1, 1, 2], dtype='uint32'),
            'c2': np.array([1.0, 2.0, 3.0], dtype='float64'),
            'c3': [True, None, False],
            'c4': ['foo', 'bar', None]
        })

        data2 = pd.DataFrame({
            'c1': np.array([3, 5], dtype='uint32'),
            'c2': np.array([4.0, 5.0], dtype='float64'),
            'c3': [True, True],
            'c4': ['baz', 'qux']
        })

        batch1 = pa.RecordBatch.from_pandas(data1)
        batch2 = pa.RecordBatch.from_pandas(data2)

>       table = pa.Table.from_batches([batch1, batch2])

pyarrow/tests/test_table.py:241:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
table.pxi:1106: in pyarrow.lib.Table.from_batches
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   pyarrow.lib.ArrowInvalid: Schema at index 1 was different:
E   c1: uint32
E   c2: double
E   c3: bool
E   c4: string
E   __index_level_0__: int64
E   -- metadata --
E   pandas: {"index_columns": ["__index_level_0__"], "column_indexes": [{"name": null, "field_name": null, "pandas_type": "unicode", "numpy_type": "object", "metadata": {"encoding": "UTF-8"}}], "columns": [{"name": "c1", "field_name": "c1
", "pandas_type": "uint32", "numpy_type": "uint32", "metadata": null}, {"name": "c2", "field_name": "c2", "pandas_type": "float64", "numpy_type": "float64", "metadata": null}, {"name": "c3", "field_name": "c3", "pandas_type": "bool", "num
py_type": "object", "metadata": null}, {"name": "c4", "field_name": "c4", "pandas_type": "unicode", "numpy_type": "object", "metadata": null}, {"name": null, "field_name": "__index_level_0__", "pandas_type": "int64", "numpy_type": "int64"
, "metadata": null}], "pandas_version": "0.23.0"}
E   vs
E   c1: uint32
E   c2: double
E   c3: bool
E   c4: string
E   __index_level_0__: int64
E   -- metadata --
E   pandas: {"index_columns": ["__index_level_0__"], "column_indexes": [{"name": null, "field_name": null, "pandas_type": "unicode", "numpy_type": "object", "metadata": {"encoding": "UTF-8"}}], "columns": [{"name": "c1", "field_name": "c1
", "pandas_type": "uint32", "numpy_type": "uint32", "metadata": null}, {"name": "c2", "field_name": "c2", "pandas_type": "float64", "numpy_type": "float64", "metadata": null}, {"name": "c3", "field_name": "c3", "pandas_type": "bool", "num
py_type": "bool", "metadata": null}, {"name": "c4", "field_name": "c4", "pandas_type": "unicode", "numpy_type": "object", "metadata": null}, {"name": null, "field_name": "__index_level_0__", "pandas_type": "int64", "numpy_type": "int64",
"metadata": null}], "pandas_version": "0.23.0"}
```

`numpy_type` values are different: bool vs object